In [1]:
#import libraries:

import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.datasets import make_regression
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [13]:
train_df = pd.read_csv('Resources/2019loans.csv')
test_df = pd.read_csv('Resources/2020Q1loans.csv')
train_df

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0           57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1          141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2          321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3           11778   11778     3000.0    0.1240       100.22           RENT   
4          169382  169382    30000.0    0.1612      1056.49       MORTGAGE   
...           ...     ...        ...       ...          ...            ...   
12175      354912  354912    19975.0    0.2565       801.09           RENT   
12176      354944  354944    15000.0    0.1774       540.34           RENT   
12177      354973  354973     3600.0    0.1862       131.28           RENT   
12178      355002  355002    15000.0    0.0881       475.68       MORTGAGE   
12179      355307  355307    15000.0    0.1774       540.34           RENT   

       annual_inc verification_status loan_status pymnt_plan  ...  \
0        223000.0        Not Verified    low_risk          n  ...   
1        123000.0     Source Verified    low_risk          n  ...   
2        197000.0     Source Verified    low_risk          n  ...   
3         45000.0        Not Verified    low_risk          n  ...   
4        133000.0     Source Verified    low_risk          n  ...   
...           ...                 ...         ...        ...  ...   
12175     28000.0        Not Verified   high_risk          n  ...   
12176     50000.0            Verified   high_risk          n  ...   
12177     60000.0        Not Verified   high_risk          n  ...   
12178     62000.0     Source Verified   high_risk          n  ...   
12179     54080.0        Not Verified   high_risk          n  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              50.0                   0.0        0.0   
1                85.0              33.3                   0.0        0.0   
2                85.7              33.3                   0.0        0.0   
3               100.0              16.7                   1.0        0.0   
4               100.0              66.7                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0             577150.0           122018.0         32000.0   
1             132750.0            27896.0         15900.0   
2             628160.0           114043.0         22600.0   
3              42006.0            20761.0         19900.0   
4             283248.0           109056.0         79500.0   
...                ...                ...             ...   
12175          50055.0            28192.0         18700.0   
12176          70324.0            57025.0         13300.0   
12177          83765.0            55156.0         14800.0   
12178         189930.0            23748.0          7000.0   
12179          48968.0            29786.0          2900.0   

      total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                       170200.0              N                     N  
1                        35398.0              N                     N  
2                        90340.0              N                     N  
3                        15406.0              N                     N  
4                        58778.0              N                     N  
...                          ...            ...                   ...  
12175                    19055.0              N                     N  
12176              

In [3]:
X_train = train_df.drop(['loan_status'], axis = 1)
X_train = pd.get_dummies(X_train)

y_train = train_df['loan_status']

In [4]:
X_test = test_df.drop(['loan_status'], axis = 1)
X_test = pd.get_dummies(X_test)

y_test = test_df['loan_status']

In [5]:
missing_cols = set( X_train.columns ) - set( X_test.columns )
# Add a missing column in test set with default value equal to 0
for c in missing_cols:
    X_test[c] = 0
# Ensure the order of column in the test set is in the same order than in train set
X_test = X_test[X_train.columns]

## Prediction : Consider the models

the machine learning algorithms Random Forest and Logistic Regression are used for various disciplines for classification and regression purposes, Logistic regression is a parameter based model and random forest is a non parametric model.
There is no single algorithm which performs best for all possible datasets,and there is no model better than the other.
For example let's take the algorithms that we will work with in this assignment, logistic regression works better when we have linearly separable data or continuous variable data, in the other hand, random forest classifier works better with Categorial variable data. Since we have a lot of categorial features in our dataset, in my opinion it is best to use the random forest classifier.

## Fit a LogisticRegression model and RandomForestClassifier model

In [6]:
clf = RandomForestClassifier(n_estimators = 100) 

clf.fit(X_train, y_train)
 
# performing predictions on the test dataset
y_pred = clf.predict(X_test)
 
# using metrics module for accuracy calculation
print("ACCURACY OF THE RANDOM FOREST CLASSIFIER MODEL: ", metrics.accuracy_score(y_test, y_pred))

ACCURACY OF THE RANDOM FOREST CLASSIFIER MODEL:  0.599532113994045


In [7]:
reg = LogisticRegression().fit(X_train, y_train)

# performing predictions on the test dataset
y_pred = reg.predict(X_test)
 
# using metrics module for accuracy calculation
print("ACCURACY OF THE LOGISTIC REGRESSION MODEL: ", metrics.accuracy_score(y_test, y_pred))

ACCURACY OF THE LOGISTIC REGRESSION MODEL:  0.5253083794130158


C:\Users\uninfor\anaconda3\envs\tensorflow\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### results and thoughts.

On unscaled data, we can see that random forest classifier performed better and got a better score.
So our prediction turned to be true

## Revisit the Preprocessing: Scale the data

In [8]:
# Scaling Our data
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Prediction : Scaled data

After scalling the data, I think that the score of Logistic Regression will improve whereas the score for Random Forest will remain the same. Graphical-model classifiers like Random Forest are invariant to feature scaling.

In [9]:
# Train the Logistic Regression model on the scaled data
logisticRegr_scaled = LogisticRegression().fit(X_train_scaled, y_train)
print("LogisticRegression scaled score: ", logisticRegr_scaled.score(X_test_scaled, y_test))

LogisticRegression scaled score:  0.7201190982560612


C:\Users\uninfor\anaconda3\envs\tensorflow\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [10]:
# Train a Random Forest Classifier model on the scaled data and print the model score
randomForestClass_scaled = RandomForestClassifier().fit(X_train_scaled, y_train)
print("RandomForestClassifier scaled score: ", randomForestClass_scaled.score(X_test_scaled, y_test))

RandomForestClassifier scaled score:  0.5946405784772437


### results and thoughts.

To conclude, scaling data greatly improved the score of the Logistic Regression model from .52 to .72 so that it outperformed the Random Forest classifier model. This shows that sometimes a simple model with scaled data can be a better fit than one with more complexity.